src/tls.c

```diff
@@ -7130,6 +7155,99 @@ static int TLSX_KeyShare_GenEccKey(WOLFSSL *ssl, KeyShareEntry* kse)
     return ret;
 }
 
+#ifdef HAVE_LIBOQS
+/* Transform a group ID into an OQS Algorithm name as a string. */
+static const char* OQS_ID2name(int id)
+{
+    switch (id) {
+        case WOLFSSL_KYBER512:        return OQS_KEM_alg_kyber_512;
+        case WOLFSSL_KYBER768:        return OQS_KEM_alg_kyber_768;
+        case WOLFSSL_KYBER1024:       return OQS_KEM_alg_kyber_1024;
+        case WOLFSSL_NTRU_HPS2048509: return OQS_KEM_alg_ntru_hps2048509;
+        case WOLFSSL_NTRU_HPS2048677: return OQS_KEM_alg_ntru_hps2048677;
+        case WOLFSSL_NTRU_HPS4096821: return OQS_KEM_alg_ntru_hps4096821;
+        case WOLFSSL_NTRU_HRSS701:    return OQS_KEM_alg_ntru_hrss701;
+        case WOLFSSL_LIGHTSABER:      return OQS_KEM_alg_saber_lightsaber;
+        case WOLFSSL_SABER:           return OQS_KEM_alg_saber_saber;
+        case WOLFSSL_FIRESABER:       return OQS_KEM_alg_saber_firesaber;
+        case WOLFSSL_KYBER90S512:     return OQS_KEM_alg_kyber_512_90s;
+        case WOLFSSL_KYBER90S768:     return OQS_KEM_alg_kyber_768_90s;
+        case WOLFSSL_KYBER90S1024:    return OQS_KEM_alg_kyber_1024_90s;
+        default:                      break;
+    }
+    return NULL;
+}
+/* Create a key share entry using liboqs parameters group.
+ * Generates a key pair.
+ *
+ * ssl   The SSL/TLS object.
+ * kse   The key share entry object.
+ * returns 0 on success, otherwise failure.
+ */
+static int TLSX_KeyShare_GenOqsKey(WOLFSSL *ssl, KeyShareEntry* kse)
+{
+    int             ret = -1;
+    const char*     algName = NULL;
+    OQS_KEM*        kem = NULL;
+    uint8_t*        pubKey = NULL;
+    uint8_t*        privKey = NULL;
+
+    algName = OQS_ID2name(kse->group);
+    if (algName == NULL) {
+        WOLFSSL_MSG("Invalid OQS algorithm specified.");
+        return BAD_FUNC_ARG;
+    }
+
+    kem = OQS_KEM_new(algName);
+    if (kem == NULL) {
+        WOLFSSL_MSG("Error creating OQS KEM, ensure algorithm support"
+                    "was enabled in liboqs.");
+        return MEMORY_E;
```

Comment:
I'm thinking that we should return `BAD_FUNC_ARG` here. It seams that the issue will more often be `an invalid algorithm name was provided, or that the requested algorithm was disabled at compile-time` than a OOM problem.

Reply:
Ah yes.  Ok.  Will do it.